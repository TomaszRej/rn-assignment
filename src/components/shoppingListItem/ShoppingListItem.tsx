import React, {FC} from 'react';
import {StyleSheet, StatusBar, Pressable} from 'react-native';
import Button from '../button/Button';
import {useAppDispatch} from '../../hooks/hooks';
import {
  IShoppingList,
  toggleArchiveList,
} from '../../store/shoppingLists/shoppingListSlice';
import {useNavigation} from '@react-navigation/native';
import Title from '../title/Title';
import {Paths} from '../../constants';

export enum ButtonTitleLabel {
  UnArchive = 'Unarchive List',
  Archive = 'Archive List',
}

interface IProps {
  item: IShoppingList;
}

const ShoppingListItem: FC<IProps> = ({item}) => {
  const dispatch = useAppDispatch();
  const navigation = useNavigation();

  const onArchiveList = () => {
    dispatch(
      toggleArchiveList({
        id: item.id,
        isArchived: item.isArchived ? false : true,
      }),
    );
  };

  const onListItemPressed = () => {
    navigation.navigate(Paths.Details, {id: item.id});
  };

  const buttonTitle = item.isArchived
    ? ButtonTitleLabel.UnArchive
    : ButtonTitleLabel.Archive;

  return (
    <Pressable
      style={styles.item}
      onPress={item.isArchived ? null : onListItemPressed}>
      <Title>{item.name}</Title>
      <Button title={buttonTitle} onPress={onArchiveList} />
    </Pressable>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginTop: StatusBar.currentHeight || 0,
  },
  item: {
    backgroundColor: '#ddd',
    padding: 20,
    marginVertical: 8,
    marginHorizontal: 16,
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 10,
  },
});

export default ShoppingListItem;
