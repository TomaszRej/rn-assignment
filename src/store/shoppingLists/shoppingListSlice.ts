import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {Sort} from '../../constants';

export interface IShoppingListDTO {
  id?: number;
  name: string;
  items: string[];
  isArchived: boolean;
  createdAt: number;
}

export interface IShoppingList extends IShoppingListDTO {
  id: number;
}

interface ShoppingListsState {
  shoppingLists: IShoppingList[];
  sortDirection: Sort;
}

const initialState: ShoppingListsState = {
  shoppingLists: [
    {
      id: 1,
      name: 'Example List',
      items: ['first item to buy', 'second item to buy'],
      isArchived: false,
      createdAt: Date.now(),
    },
  ],
  sortDirection: Sort.Desc,
};

export const shoppingListsSlice = createSlice({
  name: 'shoppingList',
  initialState,
  reducers: {
    addList: (state, action: PayloadAction<IShoppingListDTO>) => {
      state.shoppingLists = [
        ...state.shoppingLists,
        {
          ...action.payload,
          // adding fake unique id
          id: state.shoppingLists.length + 1,
        },
      ];
    },
    toggleArchiveList: (
      state,
      action: PayloadAction<{id: number; isArchived: boolean}>,
    ) => {
      state.shoppingLists = state.shoppingLists.map((list: IShoppingList) => {
        if (list.id === action.payload.id) {
          return {...list, isArchived: action.payload.isArchived};
        }
        return list;
      });
    },
    setSortDirection: (state, action: PayloadAction<Sort>) => {
      state.sortDirection = action.payload;
    },
  },
});

export const {addList, toggleArchiveList, setSortDirection} =
  shoppingListsSlice.actions;

export default shoppingListsSlice.reducer;
