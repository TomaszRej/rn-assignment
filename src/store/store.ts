import {configureStore, getDefaultMiddleware} from '@reduxjs/toolkit';
import {
  persistReducer,
  persistStore,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';
import shoppingListsSliceReducer from './shoppingLists/shoppingListSlice';

const persistConfig = {
  key: 'shoppingLists',
  storage: AsyncStorage,
  whitelist: ['shoppingLists'],
};

const persistedShoppingListsReducer = persistReducer(
  persistConfig,
  shoppingListsSliceReducer,
);

export const store = configureStore({
  reducer: {
    shoppingLists: persistedShoppingListsReducer,
  },
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
    },
  }),
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
