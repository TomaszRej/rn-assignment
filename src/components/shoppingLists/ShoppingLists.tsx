import React, {useEffect, useState} from 'react';
import {FlatList, ListRenderItem} from 'react-native';
import {useAppSelector} from '../../hooks/hooks';
import {IShoppingList} from '../../store/shoppingLists/shoppingListSlice';
import {
  getShoppingLists,
  getSortDirection,
} from '../../store/shoppingLists/selectors';
import ShoppingListItem from '../shoppingListItem/ShoppingListItem';
import {Sort} from '../../constants';

const ShoppingLists = () => {
  const shoppingLists = useAppSelector(getShoppingLists);
  const sortDirection = useAppSelector(getSortDirection);
  const [sortedList, setSortedList] = useState<IShoppingList[]>([]);

  useEffect(() => {
    const callback =
      sortDirection === Sort.Asc
        ? (a: IShoppingList, b: IShoppingList) => a.createdAt - b.createdAt
        : (a: IShoppingList, b: IShoppingList) => b.createdAt - a.createdAt;

    setSortedList(shoppingLists.sort(callback));
  }, [sortDirection, shoppingLists]);

  const renderItem: ListRenderItem<IShoppingList> = ({item}) => {
    return <ShoppingListItem item={item} />;
  };

  return <FlatList data={sortedList} renderItem={renderItem} />;
};

export default ShoppingLists;
