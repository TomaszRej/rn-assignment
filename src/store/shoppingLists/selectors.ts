import {Selector} from 'react-redux';
import type {RootState} from '../store';
import {IShoppingList} from './shoppingListSlice';
import {Sort} from '../../constants';

const getShoppingLists: Selector<RootState, IShoppingList[]> = state =>
  state.shoppingLists.shoppingLists.filter(list => !list.isArchived);

const getArchivedLists: Selector<RootState, IShoppingList[]> = state =>
  state.shoppingLists.shoppingLists.filter(list => list.isArchived);

const getShoppingListById =
  (id: number): Selector<RootState, IShoppingList | null> =>
  state => {
    const {shoppingLists} = state.shoppingLists;

    const list = shoppingLists.find(list => id === list.id);

    return list || null;
  };

const getSortDirection: Selector<RootState, Sort> = state =>
  state.shoppingLists.sortDirection;

export {
  getShoppingLists,
  getArchivedLists,
  getShoppingListById,
  getSortDirection,
};
