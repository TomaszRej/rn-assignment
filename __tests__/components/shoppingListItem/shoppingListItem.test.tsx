import React from 'react';
import {Provider} from 'react-redux';
import {store} from '../../../src/store/store';
import {render, RenderAPI} from '@testing-library/react-native';
import ShoppingListItem from '../../../src/components/shoppingListItem/ShoppingListItem';
import {IShoppingList} from '../../../src/store/shoppingLists/shoppingListSlice';
import {ButtonTitleLabel} from '../../../src/components/shoppingListItem/ShoppingListItem';

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: jest.fn(),
    }),
  };
});

describe('ShoppingListItem', () => {
  let unarchivedItem: RenderAPI;
  let archivedItem: RenderAPI;
  const unarchivedListItem: IShoppingList = {
    createdAt: 1626114143835,
    isArchived: false,
    items: [],
    id: 2,
    name: 'Test Name',
  };
  const archivedListItem: IShoppingList = {
    createdAt: 1626114143835,
    isArchived: true,
    items: [],
    id: 2,
    name: 'Test Name',
  };

  beforeEach(() => {
    unarchivedItem = render(
      <Provider store={store}>
        <ShoppingListItem item={unarchivedListItem} />,
      </Provider>,
    );
    archivedItem = render(
      <Provider store={store}>
        <ShoppingListItem item={archivedListItem} />,
      </Provider>,
    );
  });

  describe('ShoppingListItem snapshots', () => {
    it('should render correct snapshot for unarchivedListItem', () => {
      const tree = unarchivedItem.toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('should render correct snapshot for archivedListItem', () => {
      const tree = archivedItem.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('ShoppingListItem props', () => {
    it('should render correctly name', () => {
      const {getByText} = unarchivedItem;

      expect(getByText(unarchivedListItem.name)).toBeTruthy();
    });

    it('should render correct label on button', () => {
      expect.assertions(2);

      let {getByText} = unarchivedItem;

      expect(getByText(ButtonTitleLabel.Archive)).toBeTruthy();

      const {getByText: getText} = archivedItem;

      expect(getText(ButtonTitleLabel.UnArchive)).toBeTruthy();
    });
  });
});
