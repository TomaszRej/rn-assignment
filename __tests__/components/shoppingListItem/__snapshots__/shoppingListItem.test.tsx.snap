// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShoppingListItem ShoppingListItem snapshots should render correct snapshot for archivedListItem 1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ddd",
        "borderColor": "#ccc",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginHorizontal": 16,
        "marginVertical": 8,
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 32,
          "marginBottom": 20,
        }
      }
    >
      Test Name
    </Text>
    <View
      style={
        Object {
          "borderRadius": 10,
          "overflow": "hidden",
          "paddingHorizontal": 3,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#ee34dd",
                },
              ]
            }
          >
            Unarchive List
          </Text>
        </View>
      </View>
    </View>
  </View>,
  ",",
]
`;

exports[`ShoppingListItem ShoppingListItem snapshots should render correct snapshot for unarchivedListItem 1`] = `
Array [
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "#ddd",
        "borderColor": "#ccc",
        "borderRadius": 10,
        "borderWidth": 1,
        "marginHorizontal": 16,
        "marginVertical": 8,
        "padding": 20,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "black",
          "fontSize": 32,
          "marginBottom": 20,
        }
      }
    >
      Test Name
    </Text>
    <View
      style={
        Object {
          "borderRadius": 10,
          "overflow": "hidden",
          "paddingHorizontal": 3,
        }
      }
    >
      <View
        accessibilityRole="button"
        accessibilityState={Object {}}
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Array [
              Object {},
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#007AFF",
                  "fontSize": 18,
                  "margin": 8,
                  "textAlign": "center",
                },
                Object {
                  "color": "#ee34dd",
                },
              ]
            }
          >
            Archive List
          </Text>
        </View>
      </View>
    </View>
  </View>,
  ",",
]
`;
